notifications/v1/client.go

```diff
@@ -0,0 +1,158 @@
+	cancelContext context.CancelFunc
+}
+
+type SubscriberOptions struct {
+	// ProjectID is the GCP project ID that the Pub/Sub subscription belongs to. It
+	// is almost always the same GCP project that the Cloud Run service is deployed
+	// to.
+	ProjectID string
+	// SubscriptionID is the GCP Pub/Sub subscription ID to receive SAMS
+	// notifications from.
+	SubscriptionID string
+	// ReceiveSettings is the settings for receiving messages of the subscription. A
+	// zero value means to use the default settings.
+	ReceiveSettings ReceiveSettings
+	// Handlers is the collection of subscription handlers for each type of SAMS
+	// notifications.
+	Handlers SubscriberHandlers
+}
+
+func (opts SubscriberOptions) Validate() error {
+	if opts.ProjectID == "" {
+		return errors.New("ProjectID is required")
+	}
+	if opts.SubscriptionID == "" {
+		return errors.New("SubscriptionID is required")
+	}
+	return nil
+}
+
+// NewSubscriber creates a new background routine for receiving SAMS
+// notifications from given GCP project ID and Pub/Sub subscription ID.
+//
+// Users should prefer to use the top-level 'sams.NewNotificationsV1Subscriber'
+// constructor instead.
+func NewSubscriber(logger log.Logger, opts SubscriberOptions) (background.Routine, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	client, err := pubsub.NewClient(context.Background(), opts.ProjectID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
+	}
+	subscription := client.Subscription(opts.SubscriptionID)
+	subscription.ReceiveSettings = opts.ReceiveSettings
+	return &subscriber{
+		logger:       logger.Scoped("notification.subscriber"),
+		handlers:     opts.Handlers,
+		subscription: subscription,
+		state:        atomic.NewString(stateIdle),
+	}, nil
+}
+
+func (s *subscriber) Start() {
+	if !s.state.CompareAndSwap(stateIdle, stateStarted) {
+		s.logger.Error("failed to start subscriber", log.Error(errors.New(`not in "idle" state`)))
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancelContext = cancel
+	err := s.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+		var msgData struct {
+			Name     string          `json:"name"`
+			Metadata json.RawMessage `json:"metadata"`
+		}
+		err := json.Unmarshal(msg.Data, &msgData)
+		if err != nil {
+			s.logger.Error("failed to unmarshal notification message", log.Error(err))
+			msg.Nack()
+			return
+		}
+
+		err = s.handleReceive(msgData.Name, msgData.Metadata)
+		if err == nil {
+			msg.Ack()
+		} else {
+			s.logger.Error("failed to process notification message", log.Error(err))
+			msg.Nack()
+		}
+	})
+	if err != nil {
+		s.logger.Error("failed to receive notifications", log.Error(err))
+		return
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
```

Comment:
same here, IMO this should just be a panic

Reply:
or https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/files#r1595711888